src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToNewComponentCodeActionProvider.cs

```diff
@@ -100,4 +160,64 @@ private static bool TryGetNamespace(RazorCodeDocument codeDocument, [NotNullWhen
         // and causing compiler errors. Avoid offering this refactoring if we can't accurately get a
         // good namespace to extract to
         => codeDocument.TryComputeNamespace(fallbackToRootNamespace: true, out @namespace);
+
+    public (SyntaxNode? Start, SyntaxNode? End) FindContainingSiblingPair(SyntaxNode startNode, SyntaxNode endNode)
+    {
+        // Find the lowest common ancestor of both nodes
+        var lowestCommonAncestor = FindLowestCommonAncestor(startNode, endNode);
+        if (lowestCommonAncestor == null)
+        {
+            return (null, null);
+        }
+
+        SyntaxNode? startContainingNode = null;
+        SyntaxNode? endContainingNode = null;
+
+        // Pre-calculate the spans for comparison
+        var startSpan = startNode.Span;
+        var endSpan = endNode.Span;
+
+        foreach (var child in lowestCommonAncestor.ChildNodes().Where(node => node.Kind == SyntaxKind.MarkupElement))
+        {
+            var childSpan = child.Span;
+
+            if (startContainingNode == null && childSpan.Contains(startSpan))
+            {
+                startContainingNode = child;
+                if (endContainingNode != null)
+                    break; // Exit if we've found both
+            }
+
+            if (childSpan.Contains(endSpan))
+            {
+                endContainingNode = child;
+                if (startContainingNode != null)
+                    break; // Exit if we've found both
+            }
+        }
+
+        return (startContainingNode, endContainingNode);
+    }
+
+    public SyntaxNode? FindLowestCommonAncestor(SyntaxNode node1, SyntaxNode node2)
```

Comment:
Nit: This is navigating up the tree, through `node1`s parents, so it is probably better being called `FindHighestCommonAncestor`. Having said that, "ancestor" implies that anyway, so its really `FindNearestCommonAncestor`

This should be private, and hopefully static too